File: chapter1/section1.3/problem_1/use_pascal/milk_analysis3/milk.pas
program milk;

type pList = ^List;
    List = record
	farmer:longint;
	next:pList;
    end;
    HeadList = record
	head:pList;
	tail:pList;
    end;

    var fIn,fOut:text;
	sofar,i,x,want,cnt,a,b:longint;
	sorted,cost,amount:array[1..5010] of longint;
	csort:array[0..1010] of HeadList;

	t:pList;

begin
    assign(fIn,'milk.in');reset(fIn);
    assign(fOut,'milk.out'); rewrite(fOut);

    readln(fIn,want,cnt);
    for i:=1 to cnt do readln(fIn,cost[i],amount[i]);

	for i:=0 to 1000 do begin
	    new(csort[i].head);
	    csort[i].tail:=csort[i].head;
	    csort[i].head^.farmer:=-1;
	end;

	{Cast indexes into the array}
	for i:=1 to cnt do begin

	    t:=csort[cost[i]].tail;
	    if t^.farmer = -1 then t^.farmer:=i;
	    new(t^.next);
	    t^.next^.farmer:=-1;
	    csort[cost[i]].tail:=t^.next;
	end;

	{Pick indexes}
	x:=1;
	for i:=0 to 1000 do begin
	    t:=csort[i].head;
	    while t^.farmer > 0 do begin
	    sorted[x]:=t^.farmer;
	    inc(x);
	    t:=t^.next;
	end;
    end;

    sofar:=0;
    for i:=1 to cnt do begin
	if want < amount[sorted[i]] then begin
	    inc(sofar,want*cost[sorted[i]]);
	    want:=0; break;
	end

    else inc(sofar,amount[sorted[i]]*cost[sorted[i]]);
	dec(want,amount[sorted[i]]);
    end;

    writeln(fOut,sofar);
    close(fOut);
end.

File: chapter1/section1.3/problem_1/use_c/milk_analysis3_v2/milk.c
#include<stdio.h>
#include<stdlib.h>
#include<assert.h>

typedef struct List List;
typedef List *pList;
typedef struct HeadList HeadList;
struct List {
    int farmer;
    struct List * next;
};
struct HeadList {
    pList head;
    pList tail;
};

int main() {

    int cost[5011]={0};
    int amount[5011]={0};
    int sorted[5011]={0};
    int sofar,i,x,want,cnt;
    HeadList csort[1011];
    pList t;

    FILE *fin, *fout;
    fin=fopen("milk.in","r");
    fout=fopen("milk.out","w");
    assert(fin != NULL && fout != NULL);

    fscanf(fin, "%d %d", &want, &cnt);
    for (i=1;i<=cnt;i++) {
	fscanf(fin, "%d %d", &cost[i],&amount[i]);
    }
    fclose(fin);

    for (i=0;i<=1000;i++) {
	//new(csort[i].head);
	csort[i].head=(pList)malloc(sizeof (List));
	csort[i].tail=csort[i].head;
	csort[i].head->farmer=-1;
    }
    for (i=1;i<=cnt;i++) {
	t=csort[cost[i]].tail;
	if (t->farmer == -1) { t->farmer=i; }
	//new(t^.next);
	List temp;
	temp.farmer=0;
	temp.next=NULL;
	t->next=(pList)malloc(sizeof (List));

	t->next->farmer=-1;
	csort[cost[i]].tail=t->next;
    }
    x=1;
    for (i=0;i<=1000;i++) {
	t=csort[i].head;
	while (t->farmer > 0) {
	    sorted[x]=t->farmer;
	    //inc(x);
	    x++;
	    t=t->next;
	}
    }
    sofar=0;
    for (i=1;i<=cnt;i++) {
	if (want < amount[sorted[i]]) {
	    //inc(sofar,want*cost[sorted[i]]);
	    sofar += want*cost[sorted[i]];
	    want=0;
	    break;
	} else {
	    //inc(sofar,amount[sorted[i]]*cost[sorted[i]]);
	    sofar += amount[sorted[i]]*cost[sorted[i]];
	    //dec(want,amount[sorted[i]]);
	    want -= amount[sorted[i]];
	}
    }
    fprintf(fout,"%d\n",sofar);
    fclose(fout);

    return 0;

}
